Home Assistant test issue when building from source with `cython==3.0`

When building `zeroconf` from source, two tests for Home Assistant start failing.
```py
FAILED tests/components/zeroconf/test_init.py::test_setup_with_overly_long_url_and_name - TypeError: Argument 'record' has incorrect type (expected zeroconf._dns.DNSRecord, got MagicMock)
FAILED tests/components/zeroconf/test_init.py::test_setup_with_disallowed_characters_in_local_name - TypeError: Argument 'record' has incorrect type (expected zeroconf._dns.DNSRecord, got MagicMock)
```
```py
Traceback (most recent call last):
  File "/.../homeassistant/components/zeroconf/__init__.py", line 428, in _async_lookup_and_process_service_update
    await async_service_info.async_request(zeroconf, 3000)
  File "/.../site-packages/zeroconf/_services/info.py", line 677, in async_request
    out = self.generate_request_query(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/.../site-packages/zeroconf/_services/info.py", line 700, in generate_request_query
    out.add_question_or_one_cache(zc.cache, now, self.name, _TYPE_SRV, _CLASS_IN)
  File "src/zeroconf/_protocol/outgoing.py", line 186, in zeroconf._protocol.outgoing.DNSOutgoing.add_question_or_one_cache
TypeError: Argument 'record' has incorrect type (expected zeroconf._dns.DNSRecord, got MagicMock)
```

This seems to be related to the recent `3.0` for `cython`. As it isn't limited in the build dependencies
https://github.com/python-zeroconf/python-zeroconf/blob/109bbe1af010a4aecf12b5b5d697581479af88eb/pyproject.toml#L152-L154

To test it
```bash
pip uninstall -y zeroconf
pip install --no-binary zeroconf --no-cache -vv zeroconf==0.71
pytest tests/components/zeroconf/test_init.py
```

While this works
```bash
echo "cython<3.0" >> c.txt
pip uninstall -y zeroconf
PIP_CONSTRAINT=c.txt pip install --no-binary zeroconf --no-cache -vv zeroconf==0.71
pytest tests/components/zeroconf/test_init.py
```

/CC @bdraco

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1104,7 +1104,7 @@ def mock_get_source_ip() -> Generator[None, None, None]:
 @pytest.fixture
 def mock_zeroconf() -> Generator[None, None, None]:
     """Mock zeroconf."""
-    from zeroconf._cache import DNSCache  # pylint: disable=import-outside-toplevel
+    from zeroconf import DNSCache  # pylint: disable=import-outside-toplevel
 
     with patch(
         "homeassistant.components.zeroconf.HaZeroconf", autospec=True
@@ -1121,7 +1121,7 @@ def mock_zeroconf() -> Generator[None, None, None]:
 @pytest.fixture
 def mock_async_zeroconf(mock_zeroconf: None) -> Generator[None, None, None]:
     """Mock AsyncZeroconf."""
-    from zeroconf._cache import DNSCache  # pylint: disable=import-outside-toplevel
+    from zeroconf import DNSCache  # pylint: disable=import-outside-toplevel
 
     with patch("homeassistant.components.zeroconf.HaAsyncZeroconf") as mock_aiozc:
         zc = mock_aiozc.return_value

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1104,23 +1104,34 @@ def mock_get_source_ip() -> Generator[None, None, None]:
 @pytest.fixture
 def mock_zeroconf() -> Generator[None, None, None]:
     """Mock zeroconf."""
+    from zeroconf._cache import DNSCache  # pylint: disable=import-outside-toplevel
+
     with patch(
         "homeassistant.components.zeroconf.HaZeroconf", autospec=True
     ) as mock_zc, patch(
         "homeassistant.components.zeroconf.HaAsyncServiceBrowser", autospec=True
     ):
+        zc = mock_zc.return_value
+        # DNSCache has strong Cython type checks, and MagicMock does not work
+        # so we must mock the class directly
+        zc.cache = DNSCache()
         yield mock_zc
 
 
 @pytest.fixture
 def mock_async_zeroconf(mock_zeroconf: None) -> Generator[None, None, None]:
     """Mock AsyncZeroconf."""
+    from zeroconf._cache import DNSCache  # pylint: disable=import-outside-toplevel
+
     with patch("homeassistant.components.zeroconf.HaAsyncZeroconf") as mock_aiozc:
         zc = mock_aiozc.return_value
         zc.async_unregister_service = AsyncMock()
         zc.async_register_service = AsyncMock()
         zc.async_update_service = AsyncMock()
         zc.zeroconf.async_wait_for_start = AsyncMock()
+        # DNSCache has strong Cython type checks, and MagicMock does not work
+        # so we must mock the class directly
+        zc.zeroconf.cache = DNSCache()
         zc.zeroconf.done = False
         zc.async_close = AsyncMock()
         zc.ha_async_close = AsyncMock()

